tests/test_cargo_metadata.rs

```diff
@@ -0,0 +1,89 @@
+use std::fs::File;
+use std::io::prelude::*;
+
+use support::{project, execs, basic_bin_manifest};
+use hamcrest::{assert_that, existing_file, is, equal_to};
+
+
+fn setup() {
+}
+
+test!(cargo_metadata_simple {
+    let p = project("foo")
+        .file("Cargo.toml", &basic_bin_manifest("foo"));
+
+    assert_that(p.cargo_process("metadata"), execs().with_stdout(format!(r#"
+[[packages]]
+dependencies = []
+manifest_path = "{}/Cargo.toml"
+name = "foo"
+version = "0.5.0"
+
+[[packages.targets]]
+kind = ["bin"]
+name = "foo"
+src_path = "src/foo.rs"
+
+[root]
+name = "foo"
+version = "0.5.0"
+
+"#, p.root().to_str().unwrap())));
+});
+
+test!(cargo_metadata_simple_json {
+    let p = project("foo")
+            .file("Cargo.toml", &basic_bin_manifest("foo"));
+
+    assert_that(p.cargo_process("metadata").arg("-f").arg("json"), execs().with_stdout(format!(
+    "{{\"root\":{{\"name\":\"foo\",\"version\":\"0.5.0\",\"features\":null}},\"packages\":\
+[{{\"name\":\"foo\",\"version\":\"0.5.0\",\"dependencies\":[],\"targets\":[{{\"kind\":[\"bin\"],\
+\"name\":\"foo\",\"src_path\":\"src/foo.rs\",\"metadata\":null}}],\"manifest_path\":\"{}/\
+Cargo.toml\"}}]}}\n", p.root().to_str().unwrap())));
+});
+
+test!(cargo_metadata_with_invalid_manifest {
+    let p = project("foo")
+            .file("Cargo.toml", "");
+
+    assert_that(p.cargo_process("metadata"),
+    execs()
+            .with_status(101)
+            .with_stderr("\
+failed to parse manifest at `[..]`
+
+Caused by:
+  No `package` or `project` section found.
+"))
+});
+
+test!(cargo_metadata_simple_file {
+    let p = project("foo")
+            .file("Cargo.toml", &basic_bin_manifest("foo"));
```

Comment:
I think that this test may want to be a bit more flavorful. For example it could have some dependencies which need to be fetched from either git or a crates.io shim, (or just a number of path dependencies). Additionally it should have a few targets (binaries, libraries, examples, benches, etc) to get a feel for what the output will look like.